cloudinit/config/cc_ubuntu_advantage.py

```diff
@@ -247,7 +281,68 @@ def maybe_install_ua_tools(cloud):
         raise
 
 
-def handle(name, cfg, cloud, log, args):
+def _is_pro() -> bool:
+    try:
+        from uaclient.api.u.pro.attach.auto.should_auto_attach.v1 import (
+            should_auto_attach,
+        )
+    except ImportError as ex:
+        LOG.debug("Unable to import `uaclient`: %s", ex)
+        return False
+    try:
+        result = should_auto_attach()
+    except Exception as ex:
+        LOG.debug("Error during `should_auto_attach`: %s", ex)
+        LOG.warning(
+            "Unable to determine if this is an Ubuntu Pro instance."
+            " Fallback to normal UA attach."
+        )
+        return False
+    return result.should_auto_attach
+
+
+def _attach(ua_section: dict, config: Optional[dict] = None):
+    token = ua_section.get("token")
+    if not token:
+        msg = "`ubuntu-advantage.token` required in non-Pro Ubuntu instances."
+        LOG.error(msg)
+        raise RuntimeError(msg)
+    configure_ua(
+        token=token,
+        enable=ua_section.get("enable"),
+        config=config,
+    )
+
+
+def _auto_attach(ua_section: dict, config: Optional[dict] = None):
+    try:
+        from uaclient.api.u.pro.attach.auto.full_auto_attach.v1 import (
+            FullAutoAttachOptions,
+            full_auto_attach,
+        )
+        from uaclient.config import UAConfig
+    except ImportError as ex:
+        msg = f"Unable to import `uaclient`: {ex}"
+        LOG.error(msg)
+        raise RuntimeError(msg) from ex
+
+    ua_config = UAConfig(
+        **{"cfg": {"ua_config": config}} if config is not None else {}
+    )
+    options = FullAutoAttachOptions(
+        enable=ua_section.get("enable"),
+        enable_beta=ua_section.get("enable_beta"),
+        retries=_AUTO_ATTACH_RETRIES,
+    )
+    try:
+        full_auto_attach(options=options, cfg=ua_config)
+    except Exception as ex:
```

Comment:
@aciba90 I would suggest that you use exceptions from [the api exceptions file](https://github.com/canonical/ubuntu-advantage-client/blob/main/uaclient/api/exceptions.py). We are doing this to control the documented exceptions raised by the api endpoints. Incase you run to any exception that is not covered in the file, please inform me so that I can update the exceptions file.

Reply:
Done. As we do not perform any specific action depending on a concrete exception, I am catching any `UserFacingError`, which should be the parent of any exception raised here, right?

Reply:
Yes, that is correct.